config/common.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/prometheus/common/config"
+)
+
+// JoinDir returns path relative to dir.
+// If path is empty or absolute, it is returned unchanged.
+func JoinDir(dir, path string) string {
+	if path == "" || filepath.IsAbs(path) {
+		return path
+	}
+	return filepath.Join(dir, path)
+}
+
+// SetHTTPClientConfigDirectory sets filepaths to be relative to dir.
+func SetHTTPClientConfigDirectory(c *config.HTTPClientConfig, dir string) {
+	if c == nil {
+		return
+	}
+	SetTLSConfigDirectory(&c.TLSConfig, dir)
+	if a := c.BasicAuth; a != nil {
+		a.PasswordFile = JoinDir(dir, a.PasswordFile)
```

Comment:
You're modifying the user-provided config here, which is to be avoided as that could affect rendering it back and otherwise be confusing.

Reply:
This is not new. I am preserving the existing behavior.

Reply:
This is not a feature that I am trying to champion right now. I just don't want to break users that are already relying on it.

Reply:
I think it was never needed, and it's making this PR more complicated and less clean than it could be.

Reply:
@brian-brazil it is needed. This makes the relative files relative to the *config* file, instead of the working directory. That would break users.

Reply:
Files are all relative to the working directory I thought, as we never fixed that?

Reply:
https://github.com/prometheus/prometheus/blob/989260f7eddd63ad7abf14d41f70fd93af31d4c8/config/config.go#L65

Reply:
Ah, we must have fixed this back in Prometheus 2 then.

We should find a cleaner way to resolve it, such as passing it in the constructor rather than by adding whole new methods to the interface.